File: populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        
        Node* cur;
        Node* res = root;

        while(root!=NULL){
            cur = new Node(0);
            Node* base = cur;
            while(root!=NULL){
                if(root->left!=NULL){
                    cur->next = root->left;
                    cur = cur->next;
                }
                if(root->right!=NULL){
                    cur->next = root->right;
                    cur = cur->next;
                }
                root = root->next;
            }
            root = base->next;
        }
        return res;
    }
};

File: 116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':

        if root == None:
            return root
        q = Deque()
        q.append(root)
        ans = root

        while q:
            size = len(q)
            temp = Node()
            while(size):
                size -= 1
                root = q.popleft()
                temp.next = root
                temp = temp.next
                if root.left: q.append(root.left)
                if root.right: q.append(root.right)
        
        return ans

        